### PR TITLE
Add UserData properties for VM and VMSS for stable 2021-03-01

### PR DESCRIPTION
**Pull request by roipmsft**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x ] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  Late March
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  End of March
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json [hunk 1]
```diff
@@ -3247,10 +3250,11 @@
             "type": "string",
             "description": "The expand expression to apply on the operation.",
             "enum": [
-              "instanceView"
+              "instanceView",
+              "userData"
             ],
             "x-ms-enum": {
-              "name": "InstanceViewTypes",
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json:3253]: @roipmsft , changing the name here is a breaking change. Please revert this change and edit the backend if needed.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json [hunk 2]
```diff
@@ -6281,10 +6302,11 @@
...
+              "name": "ExpandTypes",
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json:6309]: Breaking change, please revert here and in the backend if necessary.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json [hunk 3]
```diff
@@ -9876,6 +9898,10 @@
           "type": "string",
           "description": "Specifies that the image or disk that is being used was licensed on-premises. <br><br> Possible values for Windows Server operating system are: <br><br> Windows_Client <br><br> Windows_Server <br><br> Possible values for Linux Server operating system are: <br><br> RHEL_BYOS (for RHEL) <br><br> SLES_BYOS (for SUSE) <br><br> For more information, see [Azure Hybrid Use Benefit for Windows Server](https://docs.microsoft.com/azure/virtual-machines/windows/hybrid-use-benefit-licensing) <br><br> [Azure Hybrid Use Benefit for Linux Server](https://docs.microsoft.com/azure/virtual-machines/linux/azure-hybrid-benefit-linux) <br><br> Minimum api-version: 2015-06-15"
         },
+        "userData": {
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json:9901]: Move this to be the last property in the list. 
This comment applies to all other times in this PR where you add a new property.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json [hunk 4]
```diff
@@ -3100,6 +3100,9 @@
           "Create a Linux vm with a patch setting patchMode of AutomaticByPlatform.": {
             "$ref": "./examples/CreateALinuxVmWithPatchSettingModeOfAutomaticByPlatform.json"
           },
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json:3102]: The title of this PR indicates these changes should be in the 2021-03-01 version. If so, please make this PR against the 2021-03-01 compute.json file and move the examples to the 2021-03-01 Examples folder.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAScaleSetWithUserData.json [hunk 5]
```diff
@@ -0,0 +1,241 @@
+{
+  "parameters": {
+    "subscriptionId": "{subscription-id}",
+    "resourceGroupName": "myResourceGroup",
+    "vmScaleSetName": "{vmss-name}",
+    "api-version": "2020-12-01",
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAScaleSetWithUserData.json:6]: API version should be 2021-03-01

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAVmWithUserData.json [hunk 6]
```diff
@@ -0,0 +1,172 @@
+{
+  "parameters": {
+    "subscriptionId": "{subscription-id}",
+    "resourceGroupName": "myResourceGroup",
+    "vmName": "myVM",
+    "api-version": "2020-12-01",
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAVmWithUserData.json:6]: 2021-03-01

**Review by Sandido — CHANGES_REQUESTED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json (hunk 4 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json:3102]: oh wow what a mistake how did i miss that

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json (hunk 1 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json]: I am still fixing the PR please dont review right now

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 7]
```diff
@@ -4349,6 +4356,20 @@
...
+            "in": "query",
+            "required": false,
+            "type": "string",
+            "description": "The expand expression to apply on the operation.",
+            "enum": [
+              "userData"
+            ],
+            "x-ms-enum": {
+              "name": "ExpandTypes",
+              "modelAsString": false
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:4371]: If you plan to add values to this enum in the future and if it can be null, I suggest setting modelAsString to 'true'.

**Review by Sandido — CHANGES_REQUESTED** [review]

**Comment by Sandido** [comment]
@roipmsft , please investigate the CI failures on this PR.

**Comment by Sandido** [comment]
You can ignore the PurchasePlan error.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 7 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json]: changed to true ty

**Comment by roipmsft** [comment]
> @roipmsft , please investigate the CI failures on this PR.

Thank you for the comments, Adam. I have resolved the comments on the PR

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json (hunk 1 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json]: reverted as you suggestedty

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json (hunk 3 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json]: done ty

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAScaleSetWithUserData.json (hunk 5 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAScaleSetWithUserData.json]: ty fixed

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAVmWithUserData.json (hunk 6 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateAVmWithUserData.json]: ty fixed

**Comment by roipmsft** [comment]
to do: Get aka.ms link from PM exlaining UserData and add to documentation

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 8]
```diff
@@ -3247,11 +3250,12 @@
...
             "description": "The expand expression to apply on the operation.",
             "enum": [
-              "instanceView"
+              "instanceView",
+              "userData"
             ],
             "x-ms-enum": {
               "name": "InstanceViewTypes",
-              "modelAsString": false
+              "modelAsString": true
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: Why is this being changed?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 9]
```diff
@@ -6281,11 +6302,12 @@
...
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:6310]: Why make this change?

**Review by Sandido — CHANGES_REQUESTED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 8 above)
Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: I wonder if this change is causing the other .NEt SDK PR check to fail. Please investigate!

**Review by Sandido — CHANGES_REQUESTED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 8 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: @Sandido I'm changing from false to true because you instructed me to do that =)

Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: @roipmsft , Hm, perhaps there was a misunderstanding. 
Enums that were previously modelAsString False should remain that way. New ones should be True.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 9 above)
Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:6310]: See above.

**Review by Sandido — CHANGES_REQUESTED** [review]

**Comment by Sandido** [comment]
@roipmsft , where is the .Net SDK PR for this change?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 8 above)
Comment (yungezz) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: change enum type name in stable API version will cause breaking change in SDKs. Pls revert the name changing part if not necessary.

**Review by yungezz — CHANGES_REQUESTED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 8 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3258]: This is necessary, unfortunately.
The way that the current existing Compute.json implements these enums is flat-out wrong.

These enums have to be correct at some point. Surely, this incorrect approach cannot remain in the Swagger forever?

I run this by the BreakingChange Review folks
https://msazure.visualstudio.com/One/_workitems/edit/9701056

**Comment by yungezz** [comment]
hi @ArcturusZhang could you pls confirm impact to GO SDK? thanks
hi @msyyc  could you pls confirm impact to Python SDK? thanks

**Comment by yungezz** [comment]
> hi @ArcturusZhang could you pls confirm impact to GO SDK? thanks
> hi @msyyc could you pls confirm impact to Python SDK? thanks

hi @ArcturusZhang could you pls confirm impact to GO SDK? thanks
hi @msyyc could you pls confirm impact to Python SDK? thanks

**Review by JeffreyRichter — APPROVED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 10]
```diff
@@ -11219,6 +11255,10 @@
         "scheduledEventsProfile": {
           "$ref": "#/definitions/ScheduledEventsProfile",
           "description": "Specifies Scheduled Event related configurations."
+        },
+        "userData": {
+          "type": "string",
+          "description": "UserData for the VMSS, which must be base-64 encoded. Customer should not pass any secrets in here. <br><br>Minimum api-version: 2021-03-01"
```
Comment (dochung4) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:11261]: It's better to spell out "VMSS" in public docs instead of using the abbreviation.

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json [hunk 11]
```diff
@@ -231,7 +231,8 @@
             "type": "string",
             "description": "The expand expression to apply on the operation.",
             "enum": [
-              "instanceView"
+              "instanceView",
+              "userData"
```
Comment (dochung4) [inline specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json:235]: Does userData need to be added here? This is under Microsoft.HybridCompute, which is a different RP I assume.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json [hunk 12]
```diff
@@ -106,7 +106,8 @@
             "description": "The expand expression to apply to the operation.",
             "type": "string",
             "enum": [
-              "instanceView"
+              "instanceView",
+              "userData"
```
Comment (dochung4) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Where are we going to document that userData is not supported for dedicated host group?

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json (hunk 11 above)
Comment (roipmsft) [reply specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json:235]: Yes, in order to add a new enum value to InstanceViewTypes (or any other existing enum), the enum value has to be added to all instances of InstanceViewTypes.
Otherwise, the PR check and AutoRestGeneration will fail with this error

“Swagger document contains two or more x-ms-enum extensions with the same name 'InstanceViewTypes' and different values: instanceView,userData vs. instanceView”

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json (hunk 12 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: @smotwani

Comment (ashintre) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Cloudservice doesn't support UserData

Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Yes, thats true. (I need to document that somewhere, but waiting for Sameer to confirm where that documentation should go)
However, in order to add a new enum value to InstanceViewTypes (or any other existing enum), the enum value has to be added to all instances of InstanceViewTypes.
Otherwise, the PR check and AutoRestGeneration will fail with this error

“Swagger document contains two or more x-ms-enum extensions with the same name 'InstanceViewTypes' and different values: instanceView,userData vs. instanceView”

**Comment by yungezz** [comment]
hi @msyyc could you pls confirm impact to Python SDK? thanks

**Review by yungezz — APPROVED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json (hunk 12 above)
Comment (ashintre) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: The name _InstanceViewTypes_ doesn't accurately describe the accepted values anymore though.

Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Yes I agree, and that is very unfortunate.
The dev who created 'InstanceViewTypes' should instead have used a naming pattern like  'ExpandTypesFor{resourceName}' (e.g. 'ExpandTypesForVirtualMachine' or 'ExpandTypesForVirtualMachineScaleSet'), which would have allowed future features to extend the $expand for different scenarios.

Unfortunately, we are now stuck with this terrible name 'InstanceViewTypes'.
Using a new enum would be a breaking change.
Since we have a work-around (adding new enum value to existing enum), BreakingChange review board says that we should use the workaround instead of breaking our customers.

Comment (smotwani) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: We should the explanation of all the enum options supported in the description of $expand  and the resources they are applicable to.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateOrUpdateAScaleSetWithUserData.json [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                  ]
+                }
+              }
+            ]
+          }
+        }
+      }
+    }
+  },
+  "responses": {
```
Comment (smotwani) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateOrUpdateAScaleSetWithUserData.json:66]: Don;t see the userData property in the response example.

File: specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json (hunk 11 above)
Comment (smotwani) [reply specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json:235]: This is different RP, we should not be making changes to the swagger file of a different RP, something is incorrect if we are expected to make changes to types of a different RP.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateOrUpdateAScaleSetWithUserData.json (hunk 13 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateOrUpdateAScaleSetWithUserData.json:66]: That is by design.
The response of create/update will never contain the UserData, to minimize the payload (this is Chango's design too).
Customer will only see UserData in response if customer explicitly does GET with $expand=userData

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json (hunk 12 above)
Comment (roipmsft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Unfortunately, CloudService Swagger and CRP Swagger are following an anti-pattern in that they both use the same enum. This means that if 'InstanceViewTypes' is changed in one service, the same change must be made in all services that also use 'InstanceViewTypes'. 

"x-ms-enum": {
              "name": "InstanceViewTypes",

In order to add a new enum value to InstanceViewTypes (or any other existing enum), the enum value has to be added to all instances of InstanceViewTypes.
Otherwise, the PR check and AutoRestGeneration will fail with this error

“Swagger document contains two or more x-ms-enum extensions with the same name 'InstanceViewTypes' and different values: instanceView,userData vs. instanceView”

The dev who created 'InstanceViewTypes' should instead have used a naming pattern like 'ExpandTypesFor{resourceName}' (e.g. 'ExpandTypesForVirtualMachine' or 'ExpandTypesForVirtualMachineScaleSet'), which would have allowed different features to extend the $expand for different scenarios. Unfortunately, we are not stuck with this 'InstanceViewTypes'. 

Any other change (such as changing ModelAsString to true, or adding an entirely new eum) is a breaking change that is very unlikely to be approved by the BreakingChange review board.

**Comment by yungezz** [comment]
hi @roipmsft , SDK generation failed with below error, could you pls have fix it? thanks
```
[AUTOREST] FATAL: System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'InstanceViewTypes' and different values: instanceView,userData vs. instanceView
```

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json (hunk 12 above)
Comment (ashintre) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/cloudService.json:110]: Okay, update the description in that case to say that userData is not supported.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 14]
```diff
@@ -11219,6 +11255,10 @@
         "scheduledEventsProfile": {
           "$ref": "#/definitions/ScheduledEventsProfile",
           "description": "Specifies Scheduled Event related configurations."
+        },
+        "userData": {
+          "type": "string",
+          "description": "UserData (which must be base-64 encoded) for the virtual machine scale set. Customer should not pass any secrets in here. <br><br>Minimum api-version: 2021-03-01"
```
Comment (ashintre) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:11261]: Keep description consistent with description of userData below.

**Review by Sandido — APPROVED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 15]
```diff
@@ -3257,9 +3262,10 @@
             "in": "query",
             "required": false,
             "type": "string",
-            "description": "The expand expression to apply on the operation.",
+            "description": "The expand expression to apply on the operation. 'InstanceView' will retrieve the instance view of the virtual machine. 'UserData' will retrieve the UserData of the virtual machine.",
```
Comment (smotwani) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:3265]: 'InstanceView' retrieves a snapshot of the runtime properties of the virtual machine that is managed by the platform and can change outside of control plane operations.
'UserData' retrieves the UserData property as part of the VM model view that was provided by the user during the VM Create/Update operation
